presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java

```diff
@@ -0,0 +1,672 @@
+     * expression is written into an existing Memo group. It validates that the
+     * the logical properties previously computed for the group are equivalent
+     * to the logical properties of the new plan that will be saved into the group.
+     * TODO Support for functional dependencies is required before we can enable this check.
+     *
+     * @return
+     */
+    public static boolean verifyMemoGroupEquivalence()
+    {
+        return false;
+    }
+
+    /**
+     * Produces the inverse mapping of the provided assignments.
+     * The inverse mapping is used to propagate individual properties across a project operation
+     * by rewriting the property's variable references to those of the
+     * output of the project operation as per the provided assignments.
+     *
+     * @param assignments
+     * @return
+     */
+    private static Map<VariableReferenceExpression, VariableReferenceExpression> inverseVariableAssignments(Assignments assignments)
+    {
+        //TODO perhaps put this in AssignmentsUtils or ProjectUtils
+        requireNonNull(assignments, "Assignments must be provided.");
+        Map<VariableReferenceExpression, VariableReferenceExpression> inverseVariableAssignments = new HashMap<>();
+        for (Map.Entry<VariableReferenceExpression, RowExpression> e : assignments.entrySet()) {
+            if (e.getValue() instanceof VariableReferenceExpression) {
+                inverseVariableAssignments.put((VariableReferenceExpression) e.getValue(), e.getKey());
+            }
+        }
+        return inverseVariableAssignments;
+    }
+
+    /**
+     * Encapsulates normalization of the key property in alignment with equivalence class property,
+     * and possible setting of max card property if a one record condition is detected.
+     * The key property is modified. Maxcard will be modified if a one record condition is detected.
+     *
+     * @param keyProperty
+     * @param maxCardProperty
+     * @param equivalenceClassProperty
+     */
+    private static void normalizeKeyPropertyAndSetMaxCard(KeyProperty keyProperty, MaxCardProperty maxCardProperty, EquivalenceClassProperty equivalenceClassProperty)
+    {
+        if (maxCardProperty.isAtMostOne()) {
+            keyProperty.empty();
+            return;
+        }
+        Optional<KeyProperty> normalizedKeyProperty = keyProperty.normalize(equivalenceClassProperty);
+        keyProperty.empty(); //add in normalized keys or set maxcard
+        if (normalizedKeyProperty.isPresent()) {
+            keyProperty.addKeys(normalizedKeyProperty.get());
+        }
+        else {
+            maxCardProperty.update(1);
+        }
+    }
+
+    /**
+     * Used by optimizer rules to determine if a set of variables is satisfied by a key constraint.
+     *
+     * @param keyVars
+     * @return
+     */
+    public boolean isDistinct(Set<VariableReferenceExpression> keyVars)
```

Comment:
`@Override`

Reply:
Added @override for all overridden methods.